File: tools/materials/schema.py
"""The substance and process gate (C9 phase 3; `SCHEMA.md`).

`table_defects()` is the sibling of this file: same shape, same posture — **print
every defect, refuse the run, quote the numbers**. What it checks is the claim the
campaign exists to make checkable: nothing is made from nothing.

The rules, and each is a rule rather than a preference because a reader has to be
able to act on it:

1. **Every substance declares a unit**, and a unit that needs a conversion declares
   it — millilitres need a density, counts need a unit mass.
2. **Everything is a rational.** A float anywhere in a quantity, density, unit mass
   or rot rate is refused, because that is precisely where a balance stops balancing
   without saying so.
3. **Every process balances in grams, exactly**, and the defect quotes both totals:
   `in 2995 g vs out 2990 g` is a sentence someone can act on and "unbalanced" is not.
4. **A process with no inputs may only produce what the world already holds** — a
   process whose output is `made` and whose inputs are empty is `grow()` wearing a
   row, and it is the one thing this table exists to refuse.
5. **Every reference resolves**: substance, tier, and any structure, tool or skill a
   process requires. The vocabularies are closed and an entry no process requires is
   refused, so they cannot fill up with intentions.
6. **A substance nothing consumes must say why**, and the reason is printed as
   **open** rather than as a defect — unfinished work with a name is not the same
   thing as a placeholder.
7. **Heat says whether it means the material or the flame.** `SOURCES.md` found the
   two readings of "kiln temperature" differ by ~600 °C and both are correct, so a
   process that declares heat without the distinction is refused.

What this gate does not decide is printed by `emit.py` alongside it: the ages, their
names and how many, the content of any rung above the stone one, and whether any of
these numbers are balanced as a game.
"""

from fractions import Fraction

import declare

#: The units a substance may be measured in, and what each one needs to reach grams.
UNITS = {
    "Mass": ("grams", None),
    "Volume": ("millilitres", "density"),
    "Count": ("units", "unit_mass"),
    "Gas": ("millilitres", "density"),
}

#: Where a substance enters the world. `made` is the one that needs a process.
SOURCES = ("mined", "grown", "gathered", "salvaged", "made")

#: The closed tag list. A substance may carry any subset; a tag outside this set is
#: a defect, because a vocabulary that grows by accident is not a vocabulary.
TAGS = ("fuel", "edible", "tool", "structure", "precious", "salvage")


def _is_rational(value) -> bool:
    return isinstance(value, (int, Fraction)) and not isinstance(value, bool)


def _grams(substance: dict, quantity) -> Fraction:
    """A declared quantity of a substance, in canonical grams, exactly.

    The conversion is the schema's own table: mass is already grams, volume
    multiplies by the declared density (g/mL), a count multiplies by the declared
    unit mass. Nothing here rounds — the return is a `Fraction` and the caller
    decides whether needing a denominator above 1 is a defect (for a *process*, it
    is: a process that produces 1/3 of a gram is a process whose numbers are wrong,
    not a process with a remainder).
    """
    unit = substance["unit"]
    if unit in ("Mass",):
        return quantity
    if unit in ("Volume", "Gas"):
        return quantity * substance["density"]
    if unit == "Count":
        return quantity * substance["unit_mass"]
    raise KeyError(f"`{unit}` is not a declared unit: {', '.join(sorted(UNITS))}")


def quantity_grams(name: str, quantity) -> Fraction:
    """The grams behind a named quantity, or a raised error naming the substance."""
    return _grams(declare.SUBSTANCES[name], quantity)


def process_totals(name: str) -> tuple:
    """What a process takes in and gives out, in grams — the number the gate quotes."""
    process = declare.PROCESSES[name]
    taken = sum((quantity_grams(sub, amount) for sub, amount in process["inputs"]), Fraction(0))
    given = sum((quantity_grams(sub, amount) for sub, amount in process["outputs"]), Fraction(0))
    return taken, given


def defects() -> list:
    """Every defect in the two tables, in one run.

    A list rather than the first: a check that reports one problem per run is a
    check somebody stops running.
    """
    found = []
    substances = declare.SUBSTANCES

    # --- the substance table -------------------------------------------------
    for name, substance in substances.items():
        where = f"substance `{name}`"
        unit = substance.get("unit")
        if unit not in UNITS:
            found.append(
                f"{where} declares unit `{unit}`, which is not one of "
                f"{', '.join(sorted(UNITS))}"
            )
            continue
        _, needs = UNITS[unit]
        if needs == "density":
            density = substance.get("density")
            if not _is_rational(density):
                found.append(
                    f"{where} is measured in {UNITS[unit][0]} and declares no rational "
                    f"density — a volume with no density cannot reach grams, so it cannot "
                    f"enter the ledger at all"
                )
            elif density <= 0:
                found.append(f"{where} declares a density of {density}, which cannot be right")
        if needs == "unit_mass":
            unit_mass = substance.get("unit_mass")
            if not _is_rational(unit_mass):
                found.append(
                    f"{where} is measured in units and declares no rational unit mass — "
                    f"\"1 hatchet\" is not a number anybody can audit"
                )
            elif unit_mass <= 0:
                found.append(f"{where} declares a unit mass of {unit_mass}")
        for field in ("density", "unit_mass", "rot"):
            value = substance.get(field)
            if value is not None and not _is_rational(value):
                found.append(
                    f"{where} declares `{field}` as {value!r}, which is not a rational — "
                    f"a float here is where the balance stops balancing"
                )
        if substance.get("source") not in SOURCES:
            found.append(
                f"{where} declares source `{substance.get('source')}`, which is not one of "
                f"{', '.join(SOURCES)}"
            )
        family = substance.get("family")
        if family not in declare.ICON_FAMILIES and family not in declare.WORLD_FAMILIES:
            found.append(f"{where} presents as `{family}`, which is not a declared family")
        hue = substance.get("hue")
        anchors = tuple(entry[0] for entry in declare.HUE_ANCHORS) + ("natural",)
        if hue not in anchors:
            found.append(f"{where} declares hue `{hue}`, which is not a declared anchor")
        for tag in substance.get("tags", ()):
            if tag not in TAGS:
                found.append(
                    f"{where} carries tag `{tag}`, which is not one of {', '.join(TAGS)}"
                )
        display = substance.get("display")
        if display is not None:
            label, factor = display
            if factor is not None and not _is_rational(factor):
                found.append(f"{where} declares a display factor of {factor!r}, which is a float")
            if not label:
                found.append(f"{where} declares a display unit with no name")

    # --- the tier table ------------------------------------------------------
    tiers = {name for name, _ordinal, _note in declare.TIERS}
    ordinals = [ordinal for _name, ordinal, _note in declare.TIERS]
    if len(set(ordinals)) != len(ordinals):
        found.append(f"two tiers share an ordinal: {ordinals}")
    if ordinals != sorted(ordinals):
        found.append(f"the tiers are not declared in order: {ordinals}")

    # --- the process table ---------------------------------------------------
    consumed = set()
    for name, process in declare.PROCESSES.items():
        where = f"process `{name}`"
        if process.get("tier") not in tiers:
            found.append(
                f"{where} belongs to tier `{process.get('tier')}`, which is not declared — "
                f"a process in no tier cannot be reached"
            )
        for side in ("inputs", "outputs"):
            for substance, quantity in process.get(side, ()):
                if substance not in substances:
                    found.append(f"{where} names `{substance}` in its {side}, which is not declared")
                    continue
                if not _is_rational(quantity):
                    found.append(f"{where} gives `{substance}` a float quantity {quantity!r}")
                    continue
                if quantity <= 0:
                    found.append(f"{where} {side} a non-positive amount of `{substance}`")
                if side == "inputs":
                    consumed.add(substance)
        mechanism = process.get("mechanism") or {}
        if not mechanism:
            found.append(
                f"{where} declares no mechanism — Q37's rule is that a mechanism is a claim "
                f"with a number, and a process without one is a name"
            )
        for field in ("hours", "labour_hours"):
            value = mechanism.get(field, Fraction(0))
            if not _is_rational(value):
                found.append(f"{where} declares `{field}` as {value!r}, which is a float")
            elif value < 0:
                found.append(f"{where} declares a negative {field}")
        if not _is_rational(mechanism.get("power_kw", 0)):
            found.append(f"{where} declares a float `power_kw`")
        heat = mechanism.get("heat_c", 0)
        if not _is_rational(heat):
            found.append(f"{where} declares heat as {heat!r}, which is a float")
        elif heat > 0 and mechanism.get("heat_kind") not in ("material", "flame"):
            found.append(
                f"{where} declares {heat} °C with no `heat_kind` — the material and the flame "
                f"differ by ~600 °C and both readings are correct (SOURCES finding 2), so a "
                f"process must say which it means"
            )
        requires = process.get("requires") or {}
        for kind, required in requires.items():
            if kind not in declare.VOCABULARY:
                found.append(
                    f"{where} requires a `{kind}`, which is not a declared vocabulary "
                    f"({', '.join(sorted(declare.VOCABULARY))})"
                )
                continue
            if required not in declare.VOCABULARY[kind]:
                found.append(
                    f"{where} requires {kind} `{required}`, which is not declared — the "
                    f"vocabulary is closed, and naming something outside it is how a process "
                    f"reaches for authority it does not have"
                )

        # The balance, quoted. This is the rule the whole file exists for.
        #
        # A process with **no inputs** is a gather, and its other side is the world
        # rather than a row: `SCHEMA.md`'s worked chain marks exactly these as
        # "source". So the balance is required of every transformation, and a gather
        # is required to say which ground it comes out of — which the extraction
        # side of the ledger is what actually checks.
        taken, given = process_totals(name)
        if process.get("inputs") and taken != given:
            found.append(
                f"{where} does not balance: in {taken} g, out {given} g — a difference of "
                f"{taken - given} g. Every gram that leaves must arrive somewhere, and a loss "
                f"is declared as an output rather than subtracted"
            )
        for total, side in ((taken, "inputs"), (given, "outputs")):
            if total.denominator != 1:
                found.append(
                    f"{where} totals {total} g across its {side}, which is not a whole number "
                    f"of grams — a process measured in thirds of a gram is a process whose "
                    f"numbers are wrong"
                )
        for substance, quantity in process.get("outputs", ()):
            if substance not in substances:
                continue
            if _grams(substances[substance], quantity).denominator != 1:
                found.append(
                    f"{where} produces {quantity} of `{substance}`, which is not a whole "
                    f"number of grams — the conversion is exact or it is a defect"
                )
        for substance, quantity in process.get("inputs", ()):
            if substance not in substances:
                continue
            if _grams(substances[substance], quantity).denominator != 1:
                found.append(
                    f"{where} consumes {quantity} of `{substance}`, which is not a whole "
                    f"number of grams"
                )

        # A gather is not a creation: no inputs means the world already held it.
        if not process.get("inputs"):
            for substance, _quantity in process.get("outputs", ()):
                declared_source = substances.get(substance, {}).get("source")
                if declared_source == "made":
                    found.append(
                        f"{where} takes nothing in and produces `{substance}`, whose source is "
                        f"`made` — that is material from nothing, which is the one thing this "
                        f"table exists to refuse. A process with no inputs may only produce "
                        f"what the world already holds"
                    )

    # --- vocabularies, and the consumer rule ---------------------------------
    for kind, entries in declare.VOCABULARY.items():
        for entry in entries:
            if not any(
                process.get("requires", {}).get(kind) == entry
                for process in declare.PROCESSES.values()
            ):
                found.append(
                    f"{kind} `{entry}` is declared and no process requires it — the vocabulary "
                    f"is closed and an entry nothing uses is an intention wearing a name"
                )

    orphaned = set()
    for name, substance in substances.items():
        if name in consumed:
            continue
        if not substance.get("no_consumer"):
            found.append(
                f"substance `{name}` has no consumer and no declared reason. The schema's rule "
                f"is that a substance nothing uses is a defect — and a substance something will "
                f"use as soon as the rung above lands is unfinished work, which must say so "
                f"with `no_consumer`"
            )
        else:
            orphaned.add(name)
    return found


def gather_from_the_world() -> list:
    """Every process that draws on the world rather than on a stock, in grams.

    Not a defect and not a balance: a gather's other side is a deposit, so what it
    produces is exactly what the ground stops holding. Listed rather than skipped,
    because "source" is the one place in the design where mass enters the world.
    """
    drawn = []
    for name, process in declare.PROCESSES.items():
        if process.get("inputs"):
            continue
        for substance, quantity in process["outputs"]:
            drawn.append((name, substance, quantity_grams(substance, quantity)))
    return drawn


def open_items() -> list:
    """What the tables deliberately do not decide, printed as open every run.

    A substance nothing consumes **is** a finding worth reading — it says which rung
    is unfinished — so it is printed here rather than swallowed. Same posture as
    `design::verify`'s open list: what a green run does *not* mean.
    """
    items = []
    for name, substance in declare.SUBSTANCES.items():
        if substance.get("no_consumer"):
            items.append(
                f"`{name}` is declared and nothing consumes it: {substance['no_consumer']}"
            )
    for name, substance, grams in gather_from_the_world():
        items.append(f"`{name}` draws {grams} g of `{substance}` out of the world's own ground")
    items.append(
        "the ages: two tiers are declared and Q69 asked for as many as possible — the "
        "content is unbounded and an age is rows, so adding one touches no schema"
    )
    items.append(
        "the numbers in these tables are declared, not measured: every duration, labour "
        "cost, power figure and loss ratio is `[D]`, and the rung above the stone one is "
        "blocked on `[NS]` source rows rather than on this schema"
    )
    return items


def main() -> int:
    # The tables are UTF-8 and the reports use em dashes, which a Windows console
    # whose code page is not UTF-8 renders as mojibake -- and a defect report nobody
    # can read is the first step to a check nobody runs. Guarded, because a stream
    # that cannot be reconfigured is not a reason to fail the gate.
    try:
        import sys
        sys.stdout.reconfigure(encoding="utf-8")
    except Exception:  # pragma: no cover - console dependent
        pass
    found = defects()
    for item in found:
        print(f"schema: DEFECT: {item}")
    for item in open_items():
        print(f"schema: open — {item}")
    if found:
        print(f"schema: {len(found)} defect(s) in {len(declare.SUBSTANCES)} substances and "
              f"{len(declare.PROCESSES)} processes")
        return 1
    print(f"schema: {len(declare.SUBSTANCES)} substances, {len(declare.PROCESSES)} processes, "
          f"{len(declare.TIERS)} tiers, all balanced")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/materials/schema.rs
//! The substance and process tables (C9 phase 3), typed, with the campaign's one
//! invariant checked where it is used.
//!
//! `tools/materials/SCHEMA.md` is the contract, `tools/materials/schema.py` is the
//! gate that refuses a row which breaks it, and this module is the runtime's view of
//! what survived. Two tables and one rule:
//!
//! * **Every substance declares a unit**, and reaches grams exactly — mass is grams,
//!   volume multiplies by a declared density, a count by a declared unit mass. There
//!   is no float in [`Rational`], and [`grams`] returns `None` rather than rounding,
//!   because a conversion that rounds is where a ledger stops balancing quietly.
//! * **Every process balances**, and [`process_totals`] reports both sides so a
//!   failure can quote them: `in 2995 g vs out 2990 g` is a sentence someone can act
//!   on. A process with **no inputs** is a gather, and [`verify`] refuses one whose
//!   output is `made` — that is `grow()` wearing a row, and it is the one thing this
//!   table exists to refuse.
//! * **The content is unbounded.** Q66 and Q69 asked for the whole substance set and
//!   as many ages as possible, so an age is rows: nothing here assumes a closed set,
//!   and adding one touches no code.
//!
//! Why the gate exists twice, in two languages: the Python one decides what may be
//! *emitted*, and this one decides what may be *used*. They are not redundant — a
//! hand-edited generated file, or a runtime path that constructs a quantity nobody
//! declared, is exactly the case the emitting gate cannot see. Mirrored on purpose,
//! the way the generator's digest is.

use crate::materials::generated::{Process, Rational, Substance};
use crate::materials::{generated, Report};

/// The canonical unit a substance is counted in.
///
/// The vocabulary is **closed**, and asking for one that is not declared is `None`
/// rather than a default — the same rule the task verbs and the case prefixes follow.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Unit {
    Mass,
    Volume,
    Count,
    Gas,
}

impl Unit {
    pub const ALL: &'static [Unit] = &[Unit::Mass, Unit::Volume, Unit::Count, Unit::Gas];

    pub fn name(self) -> &'static str {
        match self {
            Unit::Mass => "Mass",
            Unit::Volume => "Volume",
            Unit::Count => "Count",
            Unit::Gas => "Gas",
        }
    }

    /// The unit for a name, or `None` — the refusal case, not a default.
    pub fn named(name: &str) -> Option<Self> {
        Self::ALL.iter().copied().find(|unit| unit.name() == name)
    }

    /// What this unit needs before it can reach grams: a density, or a unit mass.
    pub fn needs(self) -> Option<&'static str> {
        match self {
            Unit::Volume | Unit::Gas => Some("density"),
            Unit::Count => Some("unit_mass"),
            Unit::Mass => None,
        }
    }
}

/// One substance, by its declared name.
pub fn substance(name: &str) -> Option<&'static Substance> {
    generated::SUBSTANCES.iter().find(|entry| entry.name == name)
}

/// One process, by its declared name.
pub fn process(name: &str) -> Option<&'static Process> {
    generated::PROCESSES.iter().find(|entry| entry.name == name)
}

/// The ordinal of a declared age, or `None` for a tier nobody declared — a process in
/// no tier cannot be reached, so this is the predicate a progression uses.
pub fn tier_ordinal(name: &str) -> Option<i64> {
    generated::TIERS
        .iter()
        .find(|(tier, _, _)| *tier == name)
        .map(|(_, ordinal, _)| *ordinal)
}

/// Whether a process is in an age at or below `reached`.
pub fn reachable(process: &Process, reached: i64) -> bool {
    tier_ordinal(process.tier).is_some_and(|ordinal| ordinal <= reached)
}

/// A whole number of grams, or `None` when the conversion is not exact.
///
/// `None` is the honest answer rather than a rounded one: a caller that asked for a
/// third of a gram asked a question the tables do not answer, and the Python gate
/// refuses such a row before it can ever be emitted.
pub fn grams(substance: &Substance, amount: Rational) -> Option<i64> {
    let unit = Unit::named(substance.unit)?;
    let (num, den) = match unit {
        Unit::Mass => (amount.num, amount.den),
        Unit::Volume | Unit::Gas => (
            amount.num * substance.density.num,
            amount.den * substance.density.den,
        ),
        Unit::Count => (
            amount.num * substance.unit_mass.num,
            amount.den * substance.unit_mass.den,
        ),
    };
    if den == 0 || num % den != 0 {
        return None;
    }
    Some(num / den)
}

/// A named quantity's grams, or `None` when the row is wrong — an undeclared
/// substance, an undeclared unit, or a conversion that does not land on a whole gram.
pub fn quantity_grams(name: &str, amount: Rational) -> Option<i64> {
    grams(substance(name)?, amount)
}

/// What a process takes in and gives out, in grams. `None` means some row in it could
/// not reach grams at all, which is a defect rather than a zero.
pub fn process_totals(process: &Process) -> (Option<i64>, Option<i64>) {
    let total = |side: &[(&'static str, Rational)]| {
        side.iter()
            .try_fold(0i64, |sum, (name, amount)| Some(sum + quantity_grams(name, *amount)?))
    };
    (total(process.inputs), total(process.outputs))
}

/// Whether a process balances — the one invariant the whole table exists for.
pub fn balances(process: &Process) -> bool {
    let (taken, given) = process_totals(process);
    matches!((taken, given), (Some(into), Some(out)) if into == out)
}

/// The substances that leave the world's ground rather than a stock: produced by a
/// process that takes nothing in, which is what `SCHEMA.md`'s chain marks as *source*.
///
/// Named rather than inferred at the call site, because this is the only place in the
/// design where mass legitimately enters the world — and a reader who wants to check
/// that claim should be able to ask for it directly.
pub fn gathers() -> Vec<(&'static str, &'static str, i64)> {
    let mut drawn = Vec::new();
    for process in generated::PROCESSES {
        if !process.inputs.is_empty() {
            continue;
        }
        for (name, amount) in process.outputs {
            if let Some(grams) = quantity_grams(name, *amount) {
                drawn.push((process.name, *name, grams));
            }
        }
    }
    drawn
}

/// The substance and process gate, run where the tables are used.
///
/// Reports **every** defect, so a run is one reading rather than a queue, and prints
/// the open items beside them: what a green run does *not* mean.
pub fn verify(report: &mut Report) {
    let defect = |report: &mut Report, line: String| {
        report.lines.push(format!("DEFECT: {line}"));
        report.defects.push(line);
    };

    report.lines.push(format!(
        "schema: {} substances, {} processes, {} tiers",
        generated::SUBSTANCES.len(),
        generated::PROCESSES.len(),
        generated::TIERS.len()
    ));

    // Every substance declares a unit, and a unit that needs a conversion declares it.
    for entry in generated::SUBSTANCES {
        match Unit::named(entry.unit) {
            None => defect(
                report,
                format!(
                    "substance `{}` declares unit `{}`, which is not one of {}",
                    entry.name,
                    entry.unit,
                    Unit::ALL
                        .iter()
                        .map(|unit| unit.name())
                        .collect::<Vec<_>>()
                        .join(", ")
                ),
            ),
            Some(unit) => match unit.needs() {
                Some("density") if entry.density.num <= 0 || entry.density.den <= 0 => defect(
                    report,
                    format!(
                        "substance `{}` is measured in {} and declares no positive density — \
                         it cannot reach grams, so it cannot enter the ledger",
                        entry.name,
                        unit.name().to_lowercase()
                    ),
                ),
                Some("unit_mass") if entry.unit_mass.num <= 0 || entry.unit_mass.den <= 0 => defect(
                    report,
                    format!(
                        "substance `{}` is measured in units and declares no positive unit mass \
                         — \"1 {}\" is not a number anybody can audit",
                        entry.name, entry.name
                    ),
                ),
                _ => {}
            },
        }
        if generated::ICON_FAMILIES
            .iter()
            .chain(generated::WORLD_FAMILIES.iter())
            .all(|family| family.name != entry.family)
        {
            defect(
                report,
                format!(
                    "substance `{}` presents as `{}`, which is not a declared family",
                    entry.name, entry.family
                ),
            );
        }
    }

    // Every process: an age, a balance, and references that resolve.
    let consumed: Vec<&str> = generated::PROCESSES
        .iter()
        .flat_map(|process| process.inputs.iter().map(|(name, _)| *name))
        .collect();
    for process in generated::PROCESSES {
        if tier_ordinal(process.tier).is_none() {
            defect(
                report,
                format!(
                    "process `{}` belongs to tier `{}`, which is not declared — a process in no \
                     tier cannot be reached",
                    process.name, process.tier
                ),
            );
        }
        for (name, _) in process.inputs.iter().chain(process.outputs.iter()) {
            if substance(name).is_none() {
                defect(
                    report,
                    format!("process `{}` names `{name}`, which is not declared", process.name),
                );
            }
        }
        let (taken, given) = process_totals(process);
        match (taken, given) {
            (Some(into), Some(out)) if into != out && !process.inputs.is_empty() => defect(
                report,
                format!(
                    "process `{}` does not balance: in {into} g, out {out} g — a difference of \
                     {} g. Every gram that leaves must arrive somewhere, and a loss is declared \
                     as an output rather than subtracted",
                    process.name,
                    into - out
                ),
            ),
            (None, _) | (_, None) => defect(
                report,
                format!(
                    "process `{}` has a quantity that does not land on a whole gram — a process \
                     measured in fractions of a gram is a process whose numbers are wrong",
                    process.name
                ),
            ),
            _ => {}
        }
        // A gather is not a creation: with no inputs, the world must already hold it.
        if process.inputs.is_empty() {
            for (name, _) in process.outputs {
                if substance(name).is_some_and(|entry| entry.source == "made") {
                    defect(
                        report,
                        format!(
                            "process `{}` takes nothing in and produces `{name}`, whose source is \
                             `made` — that is material from nothing, which is the one thing this \
                             table exists to refuse",
                            process.name
                        ),
                    );
                }
            }
        }
        if process.mechanism.heat_c > 0
            && !matches!(process.mechanism.heat_kind, "material" | "flame")
        {
            defect(
                report,
                format!(
                    "process `{}` declares {} °C with no heat kind — the material and the flame \
                     differ by ~600 °C and both readings are correct",
                    process.name, process.mechanism.heat_c
                ),
            );
        }
        for (kind, required) in process.requires {
            let vocabulary: &[(&str, &str)] = match *kind {
                "structure" => generated::STRUCTURES,
                "tool" => generated::TOOLS,
                "skill" => generated::SKILLS,
                other => {
                    defect(
                        report,
                        format!(
                            "process `{}` requires a `{other}`, which is not a declared \
                             vocabulary",
                            process.name
                        ),
                    );
                    continue;
                }
            };
            if !vocabulary.iter().any(|(name, _)| name == required) {
                defect(
                    report,
                    format!(
                        "process `{}` requires {kind} `{required}`, which is not declared — the \
                         vocabulary is closed, and naming something outside it is a claim with \
                         no number behind it",
                        process.name
                    ),
                );
            }
        }
    }

    // Every **made** good must be reachable from the world. This is the campaign's rule at
    // the level of a plan rather than a row: a good whose chain loops, or whose chain ends
    // in something the world does not hold, is a good nobody can produce — and the failure
    // is a cycle or a dead end in the table rather than a bad number in a row.
    for entry in generated::SUBSTANCES {
        if entry.source != "made" {
            continue;
        }
        if let Err(refusal) = crate::materials::chain::plan(entry.name, 1) {
            defect(
                report,
                format!(
                    "`{}` is made and cannot be planned from the world: {}",
                    entry.name,
                    refusal.describe()
                ),
            );
        }
    }

    // A substance nothing consumes must say why, and the reason is a finding worth
    // reading rather than a defect: it names the rung that is unfinished.
    for entry in generated::SUBSTANCES {
        if consumed.contains(&entry.name) {
            continue;
        }
        if entry.no_consumer.is_empty() {
            defect(
                report,
                format!(
                    "substance `{}` has no consumer and no declared reason — a substance \
                     something will use as soon as the rung above lands is unfinished work, \
                     which must say so",
                    entry.name
                ),
            );
        }
    }

    // The vocabularies are closed, and an entry nothing requires is an intention.
    for (kind, entries) in [
        ("structure", generated::STRUCTURES),
        ("tool", generated::TOOLS),
        ("skill", generated::SKILLS),
    ] {
        for (name, _) in entries {
            let used = generated::PROCESSES.iter().any(|process| {
                process
                    .requires
                    .iter()
                    .any(|(required_kind, required)| *required_kind == kind && required == name)
            });
            if !used {
                defect(
                    report,
                    format!(
                        "{kind} `{name}` is declared and no process requires it — an entry \
                         nothing uses is an intention wearing a name"
                    ),
                );
            }
        }
    }

    for process in generated::PROCESSES {
        let (taken, given) = process_totals(process);
        if let (Some(into), Some(out)) = (taken, given) {
            report.lines.push(format!(
                "schema: `{}` ({}) — in {into} g, out {out} g, {}/{} h of labour",
                process.name,
                process.tier,
                process.mechanism.labour_hours.num,
                process.mechanism.labour_hours.den
            ));
        }
    }
    for item in generated::SCHEMA_OPEN {
        report.lines.push(format!("schema: open — {item}"));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::materials::generated;

    /// The campaign's one invariant, checked over every row rather than over a
    /// hand-picked one: a table with one unbalanced process is a table that creates
    /// mass somewhere, and the report names which.
    #[test]
    fn every_process_balances_in_exact_grams() {
        let mut unbalanced = Vec::new();
        for process in generated::PROCESSES {
            if process.inputs.is_empty() {
                continue; // a gather draws on the world, and `gathers()` names it
            }
            let (taken, given) = process_totals(process);
            match (taken, given) {
                (Some(into), Some(out)) if into == out => {}
                (Some(into), Some(out)) => unbalanced.push(format!(
                    "`{}`: in {into} g, out {out} g, difference {} g",
                    process.name,
                    into - out
                )),
                _ => unbalanced.push(format!(
                    "`{}`: a quantity does not land on a whole gram",
                    process.name
                )),
            }
        }
        assert!(unbalanced.is_empty(), "{unbalanced:#?}");
        assert!(
            generated::PROCESSES.len() >= 7,
            "the stone rung has processes in it"
        );
    }

    /// Q7's own example, end to end: the shortest chain that ends in a tool which
    /// multiplies work, and every step of it in exact integers. `SCHEMA.md`'s worked
    /// chain is the specification; this is the machine checking it.
    #[test]
    fn the_stone_chain_builds_a_hatchet_out_of_the_world() {
        let edge = quantity_grams("knapped_edge", Rational { num: 800, den: 1 }).expect("grams");
        let haft = quantity_grams("haft_blank", Rational { num: 2400, den: 1 }).expect("grams");
        let cord = quantity_grams("cord", Rational { num: 95, den: 1 }).expect("grams");
        let hatchet =
            quantity_grams("hatchet", Rational { num: 1, den: 1 }).expect("a count reaches grams");
        assert_eq!(hatchet, 2900, "one hatchet is 2900 g, from its own unit mass");
        assert_eq!(
            edge + haft + cord,
            hatchet + 395,
            "the three inputs are the hatchet plus the declared trim waste — 3295 = 2900 + 395"
        );

        // The chain's gathers are the only mass the world gives up, and the sum is
        // what the ground stops holding.
        let drawn: i64 = gathers().iter().map(|(_, _, grams)| grams).sum();
        assert_eq!(drawn, 3000 + 1000 + 100, "the stone rung draws 4.1 kg from the world");
        assert!(
            gathers().iter().all(|(_, name, _)| substance(name)
                .is_some_and(|entry| entry.source != "made")),
            "a gather produces only what the world already holds"
        );
    }

    /// The rule that replaces `grow()`: mass does not appear. A no-input process whose
    /// output is `made` would be exactly that, so the gate must refuse it — tested by
    /// constructing the case rather than by trusting the table.
    #[test]
    fn a_process_that_makes_something_from_nothing_is_the_defect() {
        let fabricated = Process {
            name: "conjure brick",
            tier: "hands & stone",
            inputs: &[],
            outputs: &[("clay", Rational { num: 1000, den: 1 })],
            mechanism: generated::PROCESSES[0].mechanism,
            requires: &[],
            note: "",
        };
        // `clay` is `mined`, so this one is legal: the world holds it. The defect is a
        // no-input process producing something `made`, so the test asserts the rule
        // through the substance's own declared source rather than through a name.
        assert_eq!(substance("clay").expect("declared").source, "mined");
        assert!(fabricated.inputs.is_empty() && !fabricated.outputs.is_empty());
        let made = substance("haft_blank").expect("declared");
        assert_eq!(made.source, "made", "a made good must come from a process with inputs");
        for process in generated::PROCESSES {
            if process.inputs.is_empty() {
                for (name, _) in process.outputs {
                    assert_ne!(
                        substance(name).expect("declared").source,
                        "made",
                        "`{}` conjures a made good",
                        process.name
                    );
                }
            }
        }
    }

    /// A conversion that does not land on a whole gram is refused rather than rounded:
    /// rounding at the conversion is where a ledger stops balancing without saying so.
    #[test]
    fn a_conversion_that_would_round_is_refused() {
        let hatchet = substance("hatchet").expect("declared");
        assert_eq!(grams(hatchet, Rational { num: 1, den: 1 }), Some(2900));
        assert_eq!(
            grams(hatchet, Rational { num: 1, den: 3 }),
            None,
            "a third of a hatchet is not a number of grams"
        );
        assert_eq!(
            quantity_grams("nobody declared this", Rational { num: 1, den: 1 }),
            None
        );
        // Volume reaches grams through the declared density, and never through a float.
        let water = substance("water").expect("declared");
        assert_eq!(grams(water, Rational { num: 1500, den: 1 }), Some(1500), "1 mL is 1 g");
    }

    /// The unit vocabulary is closed, and each unit says what it needs.
    #[test]
    fn the_unit_list_is_closed_and_quotes_itself() {
        for unit in Unit::ALL {
            assert_eq!(Unit::named(unit.name()), Some(*unit));
        }
        assert_eq!(Unit::named("Bushel"), None);
        assert_eq!(Unit::named(""), None);
        assert_eq!(Unit::Mass.needs(), None, "grams are grams");
        assert_eq!(Unit::Volume.needs(), Some("density"));
        assert_eq!(Unit::Count.needs(), Some("unit_mass"));
    }

    /// An age is reached rather than assumed, and a process in no tier is unreachable
    /// rather than free.
    #[test]
    fn an_age_gates_the_work_inside_it() {
        assert_eq!(tier_ordinal("hands & stone"), Some(0));
        assert_eq!(tier_ordinal("bound & composite"), Some(1));
        assert_eq!(tier_ordinal("the age nobody wrote"), None);

        let bind = process("twist cord").expect("declared");
        assert!(reachable(bind, 1), "binding is reachable once its age is reached");
        assert!(!reachable(bind, 0), "and not before");
        let knap = process("knap a core").expect("declared");
        assert!(reachable(knap, 0), "the stone age is where knapping lives");
    }

    /// The gate is part of the material gate, and the open list is printed with it —
    /// a green run says the requirements are present, and nothing more.
    #[test]
    fn the_gate_reports_every_defect_and_every_open_item() {
        let report = crate::materials::verify();
        assert!(report.ok(), "the schema gate found defects: {:#?}", report.defects);
        for item in generated::SCHEMA_OPEN {
            assert!(
                report.lines.iter().any(|line| line.contains(item)),
                "the open list is part of the report: `{item}` was not printed"
            );
        }
        assert!(
            report.lines.iter().any(|line| line.contains("substances")),
            "the report states what it read"
        );
    }
}
